src/sinks/gcp/chronicle_unstructured.rs

```diff
@@ -0,0 +1,526 @@
+        let (partition_key, mut events) = input;
+        let finalizers = events.take_finalizers();
+
+        let metadata = RequestMetadata::builder(&events);
+        ((finalizers, metadata), (partition_key, events))
+    }
+
+    fn build_request(
+        &self,
+        metadata: Self::Metadata,
+        payload: EncodeResult<Self::Payload>,
+    ) -> Self::Request {
+        let (finalizers, metadata_builder) = metadata;
+
+        let metadata = metadata_builder.build(&payload);
+        let body = payload.into_payload().bytes;
+
+        ChronicleRequest {
+            body,
+            finalizers,
+            metadata,
+        }
+    }
+}
+
+impl RequestSettings {
+    fn new(config: &ChronicleUnstructuredConfig) -> crate::Result<Self> {
+        let transformer = config.encoding.transformer();
+        let serializer = config.encoding.config().build()?;
+        let encoder = crate::codecs::Encoder::<()>::new(serializer);
+        let encoder = ChronicleEncoder {
+            customer_id: config.customer_id.clone(),
+            encoder,
+            transformer,
+        };
+        Ok(Self {
+            compression: config.compression,
+            encoder,
+        })
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct ChronicleService {
+    client: HttpClient,
+    base_url: String,
+    creds: GcpAuthenticator,
+}
+
+impl ChronicleService {
+    pub const fn new(client: HttpClient, base_url: String, creds: GcpAuthenticator) -> Self {
+        Self {
+            client,
+            base_url,
+            creds,
+        }
+    }
+}
+
+impl Service<ChronicleRequest> for ChronicleService {
+    type Response = GcsResponse;
+    type Error = HttpError;
+    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;
+
+    fn poll_ready(&mut self, _: &mut std::task::Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, request: ChronicleRequest) -> Self::Future {
+        let mut builder = Request::post(&self.base_url);
+        let headers = builder.headers_mut().unwrap();
+        headers.insert(
+            "content-type",
+            HeaderValue::from_str("application/json").unwrap(),
+        );
+        headers.insert(
+            "content-length",
+            HeaderValue::from_str(&request.body.len().to_string()).unwrap(),
+        );
+
+        let mut http_request = builder.body(Body::from(request.body)).unwrap();
+        self.creds.apply(&mut http_request);
+
+        let mut client = self.client.clone();
+        Box::pin(async move {
+            let result = client.call(http_request).await;
+            result.map(|inner| GcsResponse {
+                inner,
+                protocol: "http",
+                metadata: request.metadata,
+            })
+        })
+    }
+}
+
+#[cfg(all(test, feature = "chronicle-integration-tests"))]
+mod integration_tests {
```

Comment:
Without being too familiar with new sinks additions, I noticed there seems to be one integration test that as far as I can tell is happy path only.

Are there failure paths that we could add integration tests for? And perhaps even logic in this file that is worth unit testing (without the need for setting up an integration test) ?

Reply:
Yes good call! Adding some more integration tests for non happy paths showed up some bugs - now fixed.